### DISCUSSION TITLE
Input string not in correct format

### DISCUSSION THREAD

**Discussion by techvoipit** [mcNets/mcOmron#6] [Q&A]
Hi is the first time I have tried to connect to an OMRON PLC your software is my starting point. The manufacturer gives me PLC addresses in the following format: E0.0, E10, E200.0 ....I tried to insert in the DM DM's and CI0 but it gave me the error "Input string not in the correct format" I tried to remove E letter it responded but I think values are not correct.

**Comment by mcNets** [comment]
That seems to be an address of an EXPANSION DM MEMORY AREA, I have never used it.

I've added some additional MemoryAreas: (There are 7 new memory areas depending on the bank) So you have to clone it again or pull the changes. It seems to be missing a WinForms example. You can download the example from [CodeProject](https://www.codeproject.com/Tips/878194/OMRON-PLC-TCP-Interface)

https://github.com/mcNets/mcOmron/blob/10db84a809dbf308326d45da64f1e9bda571264f/mc.omron.v1.00/FINSCommands/IFinsCommand.cs#L35-L41

So, have a look at ReadDM methods:

https://github.com/mcNets/mcOmron/blob/10db84a809dbf308326d45da64f1e9bda571264f/mc.omron.v1.00/OmronPLC.cs#L173-L184

Now, try by adding a new method, some like ReadDME:

```
public bool ReadDME(MemoryArea area, UInt16 position, ref UInt16 value)
{
	if (!finsMemoryAreadRead(area, position, 0, 1)) return false;
	value = BTool.BytesToUInt16(this._finsCmd.Response[0], this._finsCmd.Response[1]);
	return true;
}
```

Use the proper bank area: `MemoriArea.DME0, MemoriArea.DME1, ...` when you call the method and remove the 'E'.

**Comment by techvoipit** [comment]
Hi now is working but 
I don't understand how to address variables for example to use address E0.1 I have to use position 1 and for E200.1 what address have I to use ?
In the code value to set is a uint16 how I can write strings or int ? 
Thank you in advance for the help, I attach you CX Programmer screen if can help.

![image](https://github.com/mcNets/mcOmron/assets/102979164/393992dc-cb42-4c48-80ed-bdfa2be73396)

**Comment by mcNets** [comment]
You should try to find a Command reference manual, google for it, you'll find some PDF's you can download.

DM.E sounds like a Expansion Memory area. To access this memory you have to address the specific MemoryArea, it depends on your PLC model.

https://github.com/mcNets/mcOmron/blob/10db84a809dbf308326d45da64f1e9bda571264f/mc.omron.v1.00/FINSCommands/IFinsCommand.cs#L11-L44

So instead of:
```
if (!finsMemoryAreadRead(MemoryArea.DM, position, 0, 1)) return false;
```
you should try something like
```
if (!finsMemoryAreadRead(MemoryArea.DME0, position, 0, 1)) return false;
```

**Comment by techvoipit** [comment]
I made some tests and now I can write/read in all variables, I have the last thing to resolve how I can write characters in a memory area ?

**Reply by mcNets** [reply]
https://github.com/mcNets/mcOmron/blob/10db84a809dbf308326d45da64f1e9bda571264f/mc.omron.v1.00/OmronPLC.cs#L125-L145

**Reply by techvoipit** [reply]
how I can convert string in byte to write memory area?

**Reply by mcNets** [reply]
That's something that google can answer.

https://stackoverflow.com/a/16072742/3270427
